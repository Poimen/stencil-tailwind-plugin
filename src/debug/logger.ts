import chalk from 'chalk';

let _debugEnabled = false;

export function json(obj: Record<string, unknown>): void {
  _debugEnabled && console.dir(obj, { depth: null });
}

export function debug(...statements: Array<any>): void {
  _debugEnabled && console.log(chalk.bold.blue('[DBG]'), ...statements);
}

export function success(...statements: Array<any>): void {
  console.log(chalk.bold.green('[SCS]'), ...statements);
}

export function warn(...statements: Array<any>): void {
  console.log(chalk.bold.yellowBright('[WRN]'), ...statements);
}

export function error(...statements: Array<any>): void {
  console.log(chalk.bold.bgRedBright('[ERR]'), chalk.bold.red(...statements));
}

export function configureLogging(enableDebug: boolean): void {
  _debugEnabled = enableDebug;
}
