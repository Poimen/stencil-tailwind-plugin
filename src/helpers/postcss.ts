import { AcceptedPlugin } from 'postcss';
import postcssrc from 'postcss-load-config';
import cssnano from 'cssnano';
import combine from 'postcss-combine-duplicated-selectors';
import discardComments from 'postcss-discard-comments';

import { debug } from '../debug/logger';
import { PluginConfigOpts, PostcssPlugin } from '..';

async function loadPlugins(postcssConf?: string | PostcssPlugin) {
  const ctx = { plugins: [] };
  let path = '';
  if (
    typeof postcssConf === 'object' &&
    !Array.isArray(postcssConf) &&
    postcssConf !== null &&
    Array.isArray(postcssConf.plugins)
  ) {
    return postcssConf.plugins;
  } else if (typeof postcssConf === 'string' || postcssConf instanceof String) {
    path = postcssConf as string;
  }

  try {
    const configPlugins = await postcssrc(ctx as unknown, path);
    return configPlugins.plugins;
  } catch (err) {
    if (err.code === 'MODULE_NOT_FOUND') {
      throw new Error(
        `'stencil-tailwind-plugin' is not able to resolve modules required from configuration files. Make sure it is installed\nError: ${err.message}`,
      );
    }
    // No config file found, fallthrough to manually configuring postcss
    // fallthrough expected
    debug('[TW]', 'No postcss configuration file found in:', postcssConf);
  }

  return [];
}

function findIndexOfPlugin(configPlugins: postcssrc.ResultPlugin[], name: string) {
  return configPlugins.findIndex((plugin) => {
    if (typeof plugin === 'function' && plugin.name === name) {
      return true;
    }

    if (typeof plugin === 'object' && plugin !== null && (plugin['postcssPlugin'] === name || plugin['name'] === name)) {
      return true;
    }

    return false;
  });
}

interface PostcssPlugins {
  before: postcssrc.ResultPlugin[];
  after: postcssrc.ResultPlugin[];
  hasAutoPrefixer: boolean;
}

export function stripCommentsPlugin(): AcceptedPlugin {
  return discardComments({
    removeAll: true,
  });
}

export function getMinifyPlugins(): AcceptedPlugin[] {
  return [
    combine(),
    cssnano() as AcceptedPlugin,
  ];
}

export async function getPostcssPlugins(conf: PluginConfigOpts): Promise<PostcssPlugins> {
  const configPlugins = await loadPlugins(conf.postcss);

  const configPluginTailwindIdx = findIndexOfPlugin(configPlugins, 'tailwindcss');
  const hasAutoPrefixer = findIndexOfPlugin(configPlugins, 'autoprefixer') !== -1;

  const beforeTailwind = configPluginTailwindIdx === -1 ? [] : configPlugins.slice(0, configPluginTailwindIdx);
  const afterTailwind = configPluginTailwindIdx === -1 ? configPlugins : configPlugins.slice(configPluginTailwindIdx + 1);

  return {
    before: beforeTailwind,
    after: afterTailwind,
    hasAutoPrefixer,
  };
}
