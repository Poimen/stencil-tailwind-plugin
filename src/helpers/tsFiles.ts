import ts, { ModifierSyntaxKind, Node, ScriptTarget, SourceFile, SyntaxKind, TransformationContext, TransformerFactory, Transformer } from 'typescript';
import * as log from '../debug/logger';

export interface ASTMatcher {
  nodeKind: SyntaxKind;
  name?: string;
  has?: ModifierSyntaxKind;
}

export function loadTypescriptCodeFromMemory(code: string): SourceFile {
  return ts.createSourceFile('placeholder.ts', code, ScriptTarget.Latest);
}

function doesNodeMatch(walkPath: ASTMatcher[]) {
  return function(node: Node, treeLevel: number) {
    const pathNode = walkPath[treeLevel];
    if (node.kind === pathNode.nodeKind) {
      if (pathNode.has) {
        if (!node.modifiers) return false;

        let found = false;
        for (let i = 0; i < node.modifiers.length; ++i) {
          if (node.modifiers[i].kind === pathNode.has) {
            found = true;
            break;
          }
        }

        if (!found) return false;
      }

      if (pathNode.name) {
        return node['name'] && node['name'].escapedText === pathNode.name;
      }

      return true;
    }
    return false;
  };
}

export function walkTo(file: SourceFile, walkPath: ASTMatcher[]): Node {
  const doesMatch = doesNodeMatch(walkPath);
  let treeLevel = 0;

  const hasReachedLastSelector = () => treeLevel === walkPath.length;

  function walkChildNodeTree(node: Node) {
    if (doesMatch(node, treeLevel)) {
      treeLevel++;
      if (hasReachedLastSelector()) {
        return node;
      }
    }
    return ts.forEachChild(node, walkChildNodeTree);
  }
  return walkChildNodeTree(file);
}

export function makeMatcher(kind: SyntaxKind, name?: string, modifier?: ModifierSyntaxKind): ASTMatcher {
  return { nodeKind: kind, name, has: modifier };
}

export type ASTMatchCallback = (node: Node) => Node;

function getNodeToTransform(walkPath: ASTMatcher[], action: ASTMatchCallback): TransformerFactory<Node> {
  let treeLevel = 0;
  const doesMatch = doesNodeMatch(walkPath);

  const hasReachedLastSelector = () => treeLevel === walkPath.length;

  return function(context: TransformationContext): Transformer<Node> {
    function visit(node: Node): Node | Array<Node> {
      log.debug('[Typescript]', 'Visiting ' + SyntaxKind[node.kind]);

      if (doesMatch(node, treeLevel)) {
        treeLevel++;
        if (hasReachedLastSelector()) {
          return action(node);
        }

        log.debug('[Typescript]', 'Moving to next tree level', treeLevel, 'looking for', SyntaxKind[walkPath[treeLevel].nodeKind]);
        node = ts.visitEachChild(node, visit, context);
        return node;
      }

      return node;
    }

    return function(node: Node): Node {
      return ts.visitNode(node, visit);
    };
  };
}

export interface TransformedResult {
  found: boolean;
  fullText: string;
}

export function transformNodeWith(sourceFile: SourceFile, walkPath: ASTMatcher[], userAction: ASTMatchCallback): TransformedResult {
  let foundNodeInTree = false;

  function foundNodeInAST(node: Node) {
    log.debug('Found Node, calling user processing');
    foundNodeInTree = true;

    return userAction(node);
  }

  const result = ts.transform(sourceFile, [getNodeToTransform(walkPath, foundNodeInAST)]);
  const [transformed] = result.transformed;
  const printer = ts.createPrinter();
  return {
    found: foundNodeInTree,
    fullText: printer.printFile(transformed as SourceFile)
  };
}
