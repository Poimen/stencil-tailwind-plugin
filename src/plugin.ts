import { TransformResult } from 'rollup';
import PQueue from 'p-queue';
import * as log from './debug/logger';
import { transform as styleSheetTransform } from './processors/stylesheets';
import { transform as typescriptTransform } from './processors/typescript';

export function buildStart(): void {
  log.debug('Starting build');
}

export function buildEnd(err?: Error): void {
  if (err) {
    log.error('Oh noes! Something went wrong!', err.message);
    log.error(err.stack);
  }
  log.debug('Build completed');
}

function useStyleSheetTransform(filename: string) {
  return filename.match(/\.s?css/);
}

function useTypescriptXTransform(filename: string) {
  return filename.match(/\.tsx/);
}

// The build order calls the transform function before the previous transform has completed
// For functional components we have a race between completing the tsx transform and the css
// transform. Hence queue the requests until we are ready for the next file to be processed
const queue = new PQueue({ concurrency: 1 });

export async function transform(code: string, id: string): Promise<TransformResult> {
  let codeResult = code;
  if (useStyleSheetTransform(id)) {
    codeResult = await queue.add(() => styleSheetTransform(code, id));
  } else if (useTypescriptXTransform(id)) {
    codeResult = await queue.add(() => typescriptTransform(code, id));
  }

  return {
    code: codeResult,
    map: null
  };
}
