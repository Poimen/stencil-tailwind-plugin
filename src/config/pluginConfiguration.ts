import path from 'path';
import fs from 'fs-extra';
import { Config as TailwindConfig } from 'tailwindcss';
import { PluginConfigOpts, PluginConfigOptsDefaults, TailwindPluginConfig, TailwindPluginFunctionalConfig } from '../index';
import { warn } from '../debug/logger';

function makeDefaultTailwindConf(): TailwindConfig {
  return {
    content: [],
    theme: {
      extend: {}
    },
    plugins: []
  };
}

function buildTailwindConfigurationBasedOnUserObject(userConfiguration: TailwindConfig): TailwindPluginFunctionalConfig {
  return (_: string, config: TailwindConfig): TailwindConfig => {
    // eslint-disable-next-line dot-notation
    const includePreflight = config.corePlugins['preflight'];
    let preflight = includePreflight;

    if (Array.isArray(userConfiguration.corePlugins)) {
      // user configuration is disabling all core plugins...
      return {
        ...userConfiguration,
        ...config,
        corePlugins: []
      };
    }

    if (includePreflight && userConfiguration.corePlugins) {
      // if we are including the preflight, resolve to the user configuration value if available
      // eslint-disable-next-line dot-notation
      preflight = userConfiguration.corePlugins['preflight'] ?? true;
    }
    return {
      ...userConfiguration,
      content: config.content,
      corePlugins: {
        ...userConfiguration.corePlugins,
        preflight
      }
    };
  };
}

export function resolveTailwindConfigurationFromUserSettings(tailwindConf: TailwindPluginConfig, contentFileList: string[], includePreflight: boolean): TailwindConfig {
  // Resolve the Tailwind configuration based on the user function
  const tailwindResolver = tailwindConf as TailwindPluginFunctionalConfig;
  return tailwindResolver(contentFileList[0], {
    content: contentFileList,
    corePlugins: {
      preflight: includePreflight
    }
  });
}

export const PluginConfigDefaults: PluginConfigOptsDefaults = {
  DEFAULT: {
    enableDebug: false,
    tailwindCssPath: undefined,
    tailwindCssContents: '@tailwind base;@tailwind utilities;@tailwind components;',
    tailwindConf: makeDefaultTailwindConf(),
    stripComments: false,
    minify: true,
    useAutoPrefixer: true
  }
};

function fetchTailwindCssContents(tailwindCssPath?: string): string | null {
  const tailwindCssContents = fs.readFileSync(path.resolve(tailwindCssPath)).toString();

  if (tailwindCssContents.length === 0) {
    warn('No css found when reading Tailwind configuration css - path', tailwindCssPath);
    return null;
  }

  return tailwindCssContents;
}

export function configurePluginOptions(opts: PluginConfigOpts): PluginConfigOpts {
  const configuration = Object.assign({}, opts);

  if (configuration.tailwindCssPath) {
    configuration.tailwindCssContents = fetchTailwindCssContents(configuration.tailwindCssPath) ?? configuration.tailwindCssContents;
  }

  if (typeof configuration.tailwindConf !== 'function') {
    const userConfiguration = configuration.tailwindConf as TailwindConfig;
    configuration.tailwindConf = buildTailwindConfigurationBasedOnUserObject(userConfiguration);
  }

  return configuration;
}
