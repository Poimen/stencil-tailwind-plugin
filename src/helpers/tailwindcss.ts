import path from 'path';
import discardComments from 'postcss-discard-comments';
import purgecss from '@fullhuman/postcss-purgecss';
import cssnano from 'cssnano';
import tailwindcss from 'tailwindcss';
import postcss, { AcceptedPlugin } from 'postcss';
import combine from 'postcss-combine-duplicated-selectors';
import sortMediaQueries from 'postcss-sort-media-queries';
import atImport from 'postcss-import';
import autoprefixer from 'autoprefixer';
import { makeTailwindConfig, getConfiguration } from '../config/pluginConfiguration';
import { debug } from '../debug/logger';
import { PluginConfigOpts } from '../index';

function stripCommentsPlugin() {
  return discardComments({
    removeAll: true
  });
}

function getMinifyPlugins(): AcceptedPlugin[] {
  return [
    sortMediaQueries(),
    combine(),
    cssnano() as AcceptedPlugin
  ];
}

function applyRawEscaping(css: string, wasMinified: boolean): string {
  if (wasMinified) {
    // cssnano would have escaped most things...
    return css;
  }

  // cssnano was not run, do some basic escaping
  return css
    .replace(/\n/g, '')
    .replace(/'/g, '"')
    .replace(/`/g, '\\`')
    .replace(/\t/g, ' ');
}

function getPostCssPlugins(conf: PluginConfigOpts, relativePath: string, allowPurge: boolean): AcceptedPlugin[] {
  const twConf = makeTailwindConfig([relativePath]);

  const postcssPlugins: AcceptedPlugin[] = [
    atImport(conf.atImportConf) as AcceptedPlugin,
    autoprefixer(conf.autoprefixerOptions),
    tailwindcss(twConf)
  ];

  if (allowPurge && conf.enablePurge) {
    postcssPlugins.push(purgecss({
      content: [relativePath],
      safelist: conf.purgeSafeList,
      defaultExtractor: conf.purgeExtractor
    }));
  }

  if (conf.minify) {
    postcssPlugins.push(...getMinifyPlugins());
  }

  if (conf.stripComments) {
    postcssPlugins.push(stripCommentsPlugin());
  }

  return postcssPlugins;
}

export async function processSourceTextForTailwindInlineClasses(filename: string, allowPurge: boolean, sourceText?: string): Promise<string> {
  const conf = getConfiguration();
  const cssToProcess = sourceText ?? conf.tailwindCssContents;

  const relativePath = path.join('.', path.relative(process.cwd(), filename));

  const postcssPlugins = getPostCssPlugins(conf, relativePath, allowPurge);

  const result = await postcss(postcssPlugins).process(cssToProcess, { from: relativePath });

  const css = applyRawEscaping(result.css, conf.minify);

  debug('[TW]', 'Tailwind styles:', css);

  return css;
}

export async function reduceDuplicatedClassesFromFunctionalComponentInjection(filename: string, componentCss: string, injectCss: string): Promise<string> {
  const { minify, stripComments } = getConfiguration();
  const cssToProcess = componentCss + injectCss;

  if (!minify || injectCss.length === 0) {
    return cssToProcess;
  }

  const relativePath = path.join('.', path.relative(process.cwd(), filename));

  const postcssPlugins = getMinifyPlugins();

  if (stripComments) {
    postcssPlugins.push(stripCommentsPlugin());
  }
  const result = await postcss(postcssPlugins).process(cssToProcess, { from: relativePath });

  return result.css;
}
