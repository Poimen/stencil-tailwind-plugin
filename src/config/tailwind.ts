import { TailwindConfig } from 'tailwindcss/tailwind-config';

export function makeTailwindConfig(purgeFileList: string[]): TailwindConfig {
  const twConf = {
    mode: 'jit',
    purge: purgeFileList,
    darkMode: false,
    plugins: []
  };

  // HACK! Current tailwind types don't have the 'jit' mode field so make the type
  // look as if it is meant to be there
  return (twConf as unknown) as TailwindConfig;
}
