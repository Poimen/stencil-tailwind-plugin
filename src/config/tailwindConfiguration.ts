import path from 'path';
import fs from 'fs-extra';
import { TailwindConfig } from 'tailwindcss/tailwind-config';
import * as log from '../debug/logger';

let _tailwindCssContents: string;
let _tailwindConf: TailwindConfig;

export function makeTailwindConfig(purgeFileList: string[]): TailwindConfig {
  _tailwindConf.purge = purgeFileList;
  return _tailwindConf;
}

export function fetchTailwindDefaultCssConf(): string {
  return _tailwindCssContents;
}

function setDefaultTailwindCssContents(tailwindCssPath?: string) {
  const _defaultPackageTailwindCss = `
    @tailwind utilities;
    @tailwind components;
  `;

  if (tailwindCssPath) {
    _tailwindCssContents = fs.readFileSync(path.resolve(tailwindCssPath)).toString();

    if (_tailwindCssContents.length === 0) {
      log.warn('No css found when reading Tailwind configuration css - path', tailwindCssPath);
    }
  } else {
    _tailwindCssContents = _defaultPackageTailwindCss;
  }
}

function setDefaultTailwindConf(tailwindConf?: TailwindConfig) {
  if (tailwindConf) {
    _tailwindConf = tailwindConf;
  } else {
    const twConf = {
      mode: 'jit',
      purge: [],
      darkMode: false,
      plugins: []
    };
    // HACK! Current tailwind types don't have the 'jit' mode field so make the type
    // look as if it is meant to be there
    _tailwindConf = (twConf as unknown) as TailwindConfig;
  }
}

export function setupTailwindConfiguration(tailwindCssPath?: string, tailwindConf?: TailwindConfig) : void {
  setDefaultTailwindCssContents(tailwindCssPath);
  setDefaultTailwindConf(tailwindConf);
}

setDefaultTailwindConf();
setDefaultTailwindCssContents();
