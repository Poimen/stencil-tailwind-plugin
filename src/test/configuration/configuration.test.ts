import plugin, { PluginConfigOpts, PluginOpts } from '../../index';
import { getConfiguration } from '../../config/pluginConfiguration';
import { isDebugEnabled } from '../../debug/logger';
import { processSourceTextForTailwindInlineClasses } from '../../helpers/tailwindcss';
import { loadTestComponent } from '../utils';

describe('configuration', () => {
  let tailwindFrozenConfig;
  beforeAll(() => {
    const config = { conf: require('./tailwind.config') };
    tailwindFrozenConfig = Object.freeze(config.conf);
  });

  it('given configuration should set options', () => {
    // Arrange
    const opts: PluginConfigOpts = {
      enableDebug: true,
      tailwindCssPath: 'src/test/configuration/tailwind.css',
      tailwindCssContents: '',
      tailwindConf: tailwindFrozenConfig,
      stripComments: true,
      minify: false
    };
    // Act
    const result = plugin(opts);
    // Assert
    expect(getConfiguration()).toMatchSnapshot();
    expect(isDebugEnabled()).toBe(true);
    expect(result.name).toBe('tailwind');
    expect(typeof result.buildStart).toBe(typeof Function);
    expect(typeof result.buildEnd).toBe(typeof Function);
  });

  it('given *no* configuration should set default options', () => {
    // Arrange & Act
    plugin();
    // Assert
    expect(getConfiguration()).toMatchSnapshot();
    expect(isDebugEnabled()).toBe(false);
  });

  it('given modified default configuration should set default options', () => {
    // Arrange
    const opts = Object.assign({}, PluginOpts.DEFAULT, { enableDebug: false, stripComments: true });
    // Act
    plugin(opts);
    // Assert
    expect(getConfiguration()).toMatchSnapshot();
    expect(isDebugEnabled()).toBe(false);
  });

  it('given configuration that does imports to another css file, should set options', async () => {
    // Arrange
    const loadedFile = loadTestComponent('configuration', 'config-component.tsx');
    const opts: PluginConfigOpts = {
      tailwindConf: tailwindFrozenConfig,
      tailwindCssPath: 'src/test/configuration/tailwind.atimport.css',
      atImportConf: {
        path: [
          'src/test/configuration'
        ]
      }
    };
    // Act
    plugin(opts);
    // Assert
    expect(getConfiguration()).toMatchSnapshot();
    expect(await processSourceTextForTailwindInlineClasses(loadedFile.path, true, null)).toMatchSnapshot();
  });

  it('given configuration that specifies autoprefixer config, should set options', async () => {
    // Arrange
    const loadedFile = loadTestComponent('configuration', 'config-component.tsx');
    const opts: PluginConfigOpts = {
      tailwindConf: tailwindFrozenConfig,
      tailwindCssPath: 'src/test/configuration/tailwind.css',
      autoprefixerOptions: {
        grid: 'autoplace'
      }
    };
    // Act
    plugin(opts);
    // Assert
    expect(getConfiguration()).toMatchSnapshot();
  });

  it('given full postcss configuration, should override options', async () => {
    // Arrange
    const loadedFile = loadTestComponent('configuration', 'config-component.tsx');
    const opts: PluginConfigOpts = {
      tailwindConf: tailwindFrozenConfig,
      tailwindCssPath: 'src/test/configuration/tailwind.atimport.css',
      // These configuration options should not be used, set for test purposes
      autoprefixerOptions: {
        grid: 'autoplace'
      },
      // This configuration should be used alone
      postcssConfig: 'src/test/configuration/postcss.config.js'
    };
    // Act
    plugin(opts);
    // Assert
    expect(getConfiguration()).toMatchSnapshot();
    expect(await processSourceTextForTailwindInlineClasses(loadedFile.path, undefined)).toMatchSnapshot();
  });
});
