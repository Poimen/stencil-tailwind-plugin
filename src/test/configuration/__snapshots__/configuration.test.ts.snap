// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`configuration given *no* configuration should set default options 1`] = `
Object {
  "atImportConf": Object {},
  "enableDebug": false,
  "enablePurge": true,
  "minify": true,
  "purgeExtractor": [Function],
  "purgeSafeList": Array [
    ":root",
    ":host",
    ":shadow",
    "/deep/",
    "::part",
    "::theme",
  ],
  "stripComments": false,
  "tailwindConf": Object {
    "darkMode": false,
    "mode": "jit",
    "plugins": Array [],
    "purge": Array [],
  },
  "tailwindCssContents": "@tailwind base;@tailwind utilities;@tailwind components;",
  "tailwindCssPath": undefined,
}
`;

exports[`configuration given configuration should set options 1`] = `
Object {
  "atImportConf": Object {},
  "enableDebug": true,
  "enablePurge": false,
  "minify": false,
  "purgeExtractor": [Function],
  "purgeSafeList": Array [
    "Im",
    "safe",
  ],
  "stripComments": true,
  "tailwindConf": Object {
    "darkMode": false,
    "mode": "jit",
    "plugins": Array [],
    "purge": Object {
      "content": Array [
        "./src/**/*.html",
        "./src/**/*.js",
        "./src/**/*.jsx",
        "./src/**/*.tsx",
      ],
      "enabled": true,
    },
    "theme": Object {
      "extend": Object {},
    },
    "variants": Object {
      "extend": Object {},
    },
  },
  "tailwindCssContents": "@tailwind base;

@layer base {
  .autoplacement-example {
    display: grid;
    grid-template-columns: 1fr 1fr;
    grid-template-rows: auto auto;
    grid-gap: 20px;
  }
}
",
  "tailwindCssPath": "src/test/configuration/tailwind.css",
}
`;

exports[`configuration given configuration that does imports to another css file, should set options 1`] = `
Object {
  "atImportConf": Object {
    "path": Array [
      "src/test/configuration",
    ],
  },
  "enableDebug": false,
  "enablePurge": true,
  "minify": true,
  "purgeExtractor": [Function],
  "purgeSafeList": Array [
    ":root",
    ":host",
    ":shadow",
    "/deep/",
    "::part",
    "::theme",
  ],
  "stripComments": false,
  "tailwindConf": Object {
    "darkMode": false,
    "mode": "jit",
    "plugins": Array [],
    "purge": Array [],
  },
  "tailwindCssContents": "@import 'tailwind.css';
",
  "tailwindCssPath": "src/test/configuration/tailwind.atimport.css",
}
`;

exports[`configuration given configuration that does imports to another css file, should set options 2`] = `"/*! tailwindcss v2.2.9 | MIT License | https://tailwindcss.com*//*! modern-normalize v1.1.0 | MIT License | https://github.com/sindresorhus/modern-normalize */*,:after,:before{border:0 solid;box-sizing:border-box}.autoplacement-example{grid-gap:20px;display:grid;grid-template-columns:1fr 1fr;grid-template-rows:auto auto}"`;

exports[`configuration given configuration that specifies autoprefixer config, should set options 1`] = `
Object {
  "atImportConf": Object {},
  "autoprefixerOptions": Object {
    "grid": "autoplace",
  },
  "enableDebug": false,
  "enablePurge": true,
  "minify": true,
  "purgeExtractor": [Function],
  "purgeSafeList": Array [
    ":root",
    ":host",
    ":shadow",
    "/deep/",
    "::part",
    "::theme",
  ],
  "stripComments": false,
  "tailwindConf": Object {
    "darkMode": false,
    "mode": "jit",
    "plugins": Array [],
    "purge": Array [
      "src\\\\test\\\\configuration\\\\files\\\\config-component.tsx",
    ],
  },
  "tailwindCssContents": "@tailwind base;

@layer base {
  .autoplacement-example {
    display: grid;
    grid-template-columns: 1fr 1fr;
    grid-template-rows: auto auto;
    grid-gap: 20px;
  }
}
",
  "tailwindCssPath": "src/test/configuration/tailwind.css",
}
`;

exports[`configuration given configuration that specifies autoprefixer config, should set options 2`] = `"/*! tailwindcss v2.2.9 | MIT License | https://tailwindcss.com*//*! modern-normalize v1.1.0 | MIT License | https://github.com/sindresorhus/modern-normalize */*,:after,:before{border:0 solid;box-sizing:border-box}.autoplacement-example{-ms-grid-columns:1fr 20px 1fr;-ms-grid-rows:auto 20px auto;grid-gap:20px;display:-ms-grid;display:grid;grid-template-columns:1fr 1fr;grid-template-rows:auto auto}.autoplacement-example>:first-child{-ms-grid-column:1;-ms-grid-row:1}.autoplacement-example>:nth-child(2){-ms-grid-column:3;-ms-grid-row:1}.autoplacement-example>:nth-child(3){-ms-grid-column:1;-ms-grid-row:3}.autoplacement-example>:nth-child(4){-ms-grid-column:3;-ms-grid-row:3}"`;

exports[`configuration given modified default configuration should set default options 1`] = `
Object {
  "atImportConf": Object {},
  "enableDebug": false,
  "enablePurge": true,
  "minify": true,
  "purgeExtractor": [Function],
  "purgeSafeList": Array [
    ":root",
    ":host",
    ":shadow",
    "/deep/",
    "::part",
    "::theme",
  ],
  "stripComments": true,
  "tailwindConf": Object {
    "darkMode": false,
    "mode": "jit",
    "plugins": Array [],
    "purge": Array [],
  },
  "tailwindCssContents": "@tailwind base;@tailwind utilities;@tailwind components;",
  "tailwindCssPath": undefined,
}
`;
